futures-util/src/await/join.rs

```diff
@@ -0,0 +1,38 @@
+//! The `join` macro.
+
+#[macro_export]
+macro_rules! join {
+    ($($fut:ident),*) => { {
+        $(
+            let mut $fut = $crate::future::maybe_done($fut);
+            pin_mut!($fut);
+        )*
+        loop {
+            let mut all_done = true;
+            $(
+                if let ::core::task::Poll::Pending = poll!($fut.reborrow()) {
```

Comment:
Currently this and all other macros in the `futures` crate assume that `::core` is available-- we should make all the macro-containing `futures` crates re-export `core`, or at least `core::{task, future}` and `PInMut` so that the macros can use `$crate::core_reexport::...` cc @MajorBreakfast @aturon

Reply:
Yes this makes sense. Without this we just get `failed to resolve. Maybe a missing extern crate core;?` errors.

Reply:
Yeah-- not sure what happens in the new edition.

Reply:
Does `#[doc(hidden)]` work for modules?

Reply:
Yeah, presumably it would be a `#[doc(hidden)]` reexport of `core`.